compiler/rustc_const_eval/src/check_consts/qualifs.rs

```diff
@@ -347,18 +347,21 @@ where
 
     // Check the qualifs of the value of `const` items.
     let uneval = match constant.const_ {
-        Const::Ty(_, ct)
-            if matches!(
-                ct.kind(),
-                ty::ConstKind::Param(_) | ty::ConstKind::Error(_) | ty::ConstKind::Value(_)
-            ) =>
-        {
-            None
-        }
-        Const::Ty(_, c) => {
-            bug!("expected ConstKind::Param or ConstKind::Value here, found {:?}", c)
-        }
+        // Type-level constants that are already concrete or are params/errors: fall back to
+        // type-based qualifs below.
+        Const::Ty(_, ct) => match ct.kind() {
+            ty::ConstKind::Param(_) | ty::ConstKind::Error(_) | ty::ConstKind::Value(_) => None,
+            // A type-level unevaluated const (e.g., from generic const exprs). Treat like a
+            // MIR unevaluated const so we can look up qualifs on the referenced def.
+            ty::ConstKind::Unevaluated(uv) => Some(mir::UnevaluatedConst::new(uv.def, uv.args)),
```

Comment:
the fact that we've got an unevaluated const ty as a constant value in the body is wrong :thinking: we should not support this but instead fix [MIR building for] GCE to not emit such a thing.

Generally, we're currently working towards a replacement of `generic_const_exprs` which won't have these issues, so we're currently not very actively trying to fix bugs of that feature

Reply:
Thank you.